### PR TITLE
Kaa-1172: Add a page on Administration guide -> System components overview

### PR DESCRIPTION
**Pull request by ashvayka**

### CODE REVIEW

File: doc/Administration-guide/System-components-overview/index.md [hunk 1]
```diff
@@ -5,63 +5,73 @@ permalink: /:path/
...
-    * [Control service](#control-service)
-    * [Operations service](#operations-service)
-    * [Bootstrap service](#bootstrap-service)
-  * [Endpoint](#endpoint)
-  * [Third-party](#third-party)
-    * [Zookeeper](#zookeeper)
-    * [SQL database](#sql-database)
-    * [NoSQL database](#nosql-database)
-* [Further reading](#further-reading)
+The Kaa IoT platform consists of the Kaa server, Kaa extensions, and endpoint SDKs (see [Architecture overview]({{root_url}}Architecture-overview/) for more details).
```
Comment (balldir) [inline doc/Administration-guide/System-components-overview/index.md]: According to our "Documentation contribution and style guide" something like "This guide will provide the overview of key system components from the administration point of view" is required.

File: doc/Administration-guide/System-components-overview/index.md [hunk 2]
```diff
@@ -5,63 +5,73 @@ permalink: /:path/
...
-### NoSQL database ###
-NoSQL database instance is used to store information about endpoint profiles, notifications, configurations, etc. The volume of this information scales linearly with amount of endpoints that are managed using particular Kaa cluster instance. NoSQL database nodes can be co-located with Kaa nodes on the same physical or virtual machines. Kaa support Apache Cassandra and MongoDB as a NoSQL database at the moment.
+## Scalability
+
+Multiple nodes with Operations service enabled may be set up in a Kaa cluster for the purpose of horizontal scaling of endpoints requests processing. 
+Multiple nodes with Control service enabled provide horizontal scalability of administration REST API calls processing.
+
+## Load Balancing (LB)
+
+LB task can be decoupled into two subtasks based on originator of requests to Kaa cluster.
```
Comment (balldir) [inline doc/Administration-guide/System-components-overview/index.md]: After reading this sentence I have no idea what these tasks are. I think they should be spelled inline.

File: doc/Administration-guide/System-components-overview/index.md [hunk 3]
```diff
@@ -5,63 +5,73 @@ permalink: /:path/
...
-A tenant in Kaa is a separate business entity which contains its own endpoints, applications and users.
-A single Kaa deployment is able to support multiple tenants, with multiple applications per each tenant. As illustrated in the following diagram, applications belong to tenants, while endpoints register within applications. In addition to tenants, applications and endpoints, there are also users and endpoint groups, which will be described in detail further in this reference.
+Each Kaa node in cluster runs combination of Control, Operations, and Bootstrap services. 
+A system administrator can enable or disable any of the services on a particular node, thus flexibly configuring the cluster deployment (see [general configuration]({{root_url}}Administration-guide/System-Configuration/General-configuration/) for more details).
 
-![Kaa instance (cluster)](1-04.png "Kaa instance (cluster)")
+Kaa Control service is responsible for managing overall system data, processing API calls from the web UI and external integrated systems, and delivering corresponding notifications to Operations services.
```
Comment (ikulikov) [inline doc/Administration-guide/System-components-overview/index.md]: 'processing API calls from the web UI and external integrated systems' - 'processing API calls from the Administration UI and external integrated systems communicating via REST API'

Comment (ikulikov) [inline doc/Administration-guide/System-components-overview/index.md]: I think we should mention also that Control service embeds Administration UI web server.

File: doc/Administration-guide/System-components-overview/index.md [hunk 4]
```diff
@@ -5,63 +5,73 @@ permalink: /:path/
...
-Multiple nodes with Operations service enabled may be set up in a Kaa cluster for the purpose of horizontal scaling. In this case, all the Operations service will function concurrently. In case an Operations service outage happens, the corresponding endpoints switch to the other available Operations services automatically. A Kaa cluster provides instruments for the workload re-balancing at run time, thus effectively routing endpoints to the less loaded nodes in the cluster.
+#### SQL database
 
-### Bootstrap service ###
-A Kaa Bootstrap service is responsible for directing endpoints to Operations services. On their part, Kaa endpoints have a built-in list of Bootstrap services set up in the given Kaa deployment. The endpoints use this list to query the Bootstrap services and retrieve a list of currently available Operations services from them, as well as security credentials. Bootstrap services maintain their lists of available Operations service nodes by coordinating with the ZooKeeper servers.
+SQL database instance is used to store tenants, applications, endpoint groups and other metadata that does not grow with the increase in the amount of endpoints.
```
Comment (ikulikov) [inline doc/Administration-guide/System-components-overview/index.md:42]: and other metadata that is not related to particular endpoints so its size does not depend on the amount of managed endpoints.

File: doc/Administration-guide/System-components-overview/index.md [hunk 5]
```diff
@@ -5,63 +5,73 @@ permalink: /:path/
...
-Apache ZooKeeper enables highly reliable distributed coordination of Kaa cluster nodes. Each Kaa node push information about connection parameters, enabled services and corresponding services load. Other Kaa nodes use this information in order to get list of their neighbors and communicate with them. Control service uses information about existing Bootstrap services and their connection parameters during SDK generation.
+## High availability (HA)
 
-### SQL database ###
-SQL database instance is used to store metadata about tenants, applications, endpoint groups, etc. This information is shared between endpoints, thus it's volume does not scale and it can be efficiently stored in modern SQL databases. To support high availability of Kaa cluster, SQL database should be also deployed in cluster mode.
+High availability of a Kaa cluster is achieved by deploying both SQL and NoSQL databases in HA mode. 
+Database nodes can be co-located with Kaa nodes on the same physical or virtual machines.
+In order to provide HA of Kaa services, one should deploy topology with at least two Kaa nodes. Each node should act as Control, Operations and Boostrap service simultaniously.
```
Comment (ikulikov) [inline doc/Administration-guide/System-components-overview/index.md]: It is not clear how HA works in general. As I understand currently each node has same weight and acts independently. Please add some words about this.

File: doc/Administration-guide/System-components-overview/index.md [hunk 6]
```diff
@@ -5,63 +5,73 @@ permalink: /:path/
...
+
+Kaa SDK choose Bootstrap and Operations service instances pseudo-randomly during session initiation.
+However, under the conditions when the cluster is heavily loaded, random endpoints distribution may not be good enough.
+Further, as a node joins the cluster, there is a need to re-balance the load across the new cluster set up.
+Kaa server uses the active LB approach to instruct some of the endpoints to reconnect to a different Operations service, thus equalizing the load across the nodes.
+The algorithm takes servers load information (connected endpoints count, load average, etc.) published by Kaa nodes as an input, and periodically recalculates weights of each node.
+Further, the overloaded nodes are instructed to redirect some of the connecting endpoints to a different node.
+
+A similar approach can be used for offloading all of the load from a node subject to a scheduled service, or to gradually migrate the cluster across the physical or virtual machines.
+Doing so requires setting of a custom LB strategy.
```
Comment (ikulikov) [inline doc/Administration-guide/System-components-overview/index.md]: Can you provide any reference about how to configure custom LB strategy?

**Comment by balldir** [comment]
LGTM

File: doc/Administration-guide/System-components-overview/index.md [hunk 7]
```diff
@@ -5,63 +5,82 @@ permalink: /:path/
...
+Kaa officially supports [Apache Cassandra](http://cassandra.apache.org/) and [MongoDB](https://www.mongodb.com/) as embedded NoSQL database at the moment.
 
-### SQL database ###
-SQL database instance is used to store metadata about tenants, applications, endpoint groups, etc. This information is shared between endpoints, thus it's volume does not scale and it can be efficiently stored in modern SQL databases. To support high availability of Kaa cluster, SQL database should be also deployed in cluster mode.
+## High availability (HA)
 
-### NoSQL database ###
-NoSQL database instance is used to store information about endpoint profiles, notifications, configurations, etc. The volume of this information scales linearly with amount of endpoints that are managed using particular Kaa cluster instance. NoSQL database nodes can be co-located with Kaa nodes on the same physical or virtual machines. Kaa support Apache Cassandra and MongoDB as a NoSQL database at the moment.
+In order to provide HA of Kaa services, one should deploy topology with at least two Kaa nodes. Each node should act as Control, Operations and Boostrap service simultaniously.
```
Comment (Polinaa) [inline doc/Administration-guide/System-components-overview/index.md:52]: Boo**t**strap service simultan**e**ously.

File: doc/Administration-guide/System-components-overview/index.md [hunk 8]
```diff
@@ -5,63 +5,82 @@ permalink: /:path/
...
+Multiple nodes with Operations service enabled may be set up in a Kaa cluster for the purpose of horizontal scaling of endpoints requests processing. 
+Multiple nodes with Control service enabled provide horizontal scalability of administration REST API calls processing.
+
+## Load Balancing (LB)
+
+LB task can be decoupled into two subtasks based on originator of requests to Kaa cluster: Kaa Endpoint SDK and REST API requests.
+
+#### Kaa Endpoint SDK requests
+
+Kaa SDK choose Bootstrap and Operations service instances pseudo-randomly during session initiation.
```
Comment (Polinaa) [inline doc/Administration-guide/System-components-overview/index.md:72]: choose**s**

File: doc/Administration-guide/System-components-overview/index.md [hunk 9]
```diff
@@ -5,63 +5,82 @@ permalink: /:path/
 sort_idx: 10
 ---
 
-Guide for the platform administrators.
+{% assign root_url = page.url | split: '/'%}
```
Comment (Polinaa) [inline doc/Administration-guide/System-components-overview/index.md:8]: Include variables.md files to capture root_url.
